community/rfcs/25-23-04-GenAIComps-001-Air-Gap-Support.md

```diff
@@ -0,0 +1,81 @@
+
+In this RFC, we'll mainly cover the `Type2` and `Type3` online data type, since most of the microservices have already supported using the offline data of `Type1`. 
+
+## Motivation
+
+When trying to deploy the OPEA microservices in some customers air-gapped environment, we've found that there are quite some OPEA microservices needs to download data from internet during runtime, which requires special tweaks for each of them. We want to make sure that all OPEA microservices can be run in the air-gapped environment in a uniform way.
+
+## Design Proposal
+
+### Ways to verify
+To quickly verify if a OPEA microservice support air-gapped mode or not, and what kind of online data it's downloading, we can use the following steps:
+
+ 1. Deploy the microservice in one of the 3 following environment:
+    - Deploy it in a real air gapped environment(docker or K8s)
+    - Deploy it in a tweak K8s environment by disable K8s DNS forwarding(i.e. remove the `forward` part in `kubectl -n kube-system edit cm coredns`, and restart coredns related pods)
+    - Deploy it with the fake proxy settings by setting both `http_proxy` and `https_proxy` to a non existent proxy servers, e.g. `http://localhost:54321`.
+ 2. Send requests to the microservice under verification
+     We need to make sure that all the sent out requests should have a decent coverage of the internal data flow of that microservices, because sometimes it's not the microservice itself is downloading online data, but the dependent modules.
+ 3. Check the requests return status and microservice logs to find out whether it supports air-gapped mode and what kind of online data it's downloading if it's not.
+
+### Type 2 online data
+
+Since we're not allowed the distribute AI model data in the microserivce's container image, we need to make sure that there is a way to `mount` the user pre downloaded AI model data into the microservice's runtime so that the microservice itself can run in the air gapped mode. We also need to document what to be download and how to `mount` in the deployment document of that microservice.
+
+### Type 3 online data
+
+To minimize the deployment complexity, for small size online data which is NOT shared by multiple microservices, we should have them downloaded in the container image, so that the microservice itself doesn't need to download it during runtime.
+
+For online data which are used by multiple microservices, e.g. `nltk` data,  depends on its size, we can have them pre downloaded in the container image if it will not increase the image size more than 5%. For large size data, we should follow the `Type 2` online data method to support running in air gapped mode.
```

Comment:
I remember that `nltk` is already downloaded in the Docker image [HERE[(https://github.com/opea-project/GenAIComps/blob/main/comps/third_parties/gpt-sovits/src/Dockerfile#L30). Is there anywhere else that download the `nltk` package?

Reply:
It seems that the current nltk downloading is not enough.